tools/ncnnoptimize.cpp

```diff
@@ -104,8 +128,184 @@ class NetOptimize : public ncnn::Net
+    int support_fp32_conv_type(ncnn::Convolution* op, ncnn::Mat& mat, int type);
 };
 
+void NetOptimize::gauss_random(ncnn::Mat &m)
+{
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    std::normal_distribution<float> d(1.0f, 1.0f);
+    int size = m.c * m.h * m.w;
+    for (int i = 0; i < size; ++i)
+    {
+        (static_cast<float*>(m.data))[i] = d(gen);
+    }
+}
+
+
+void NetOptimize::find_fastest_fp32_conv(const char* dataname, int w, int h, int c)
+{
+    ncnn::PoolAllocator allocator;
+    allocator.clear();
+
+    ncnn::Option opt;
+    // embeded system generally use single thread
+    opt.num_threads = 1;
+
+    const int layer_count = layers.size();
+    ncnn::Extractor ex = create_extractor();
+    
+    ncnn::Mat input(w, h, c);
+    if (ex.input(dataname, input) < 0)
+    {
+        fprintf(stderr, "set input failed, check dataname.\n");
+        return;
+    }
+
+    const char* IMPL_NAME[6] = {"baseline", "winograd", "pointwise", "im2col", "direct", "conv3x3s2"};
+    for (int i = 0; i < layer_count; ++i)
+    {
+        ncnn::Layer* layer = layers[i];
+        if (layer->type == "Convolution")
+        {
+            ncnn::Convolution* op = (ncnn::Convolution*)layer;
+
+            ncnn::Mat bottom_blob;
+            ncnn::Mat top_blob;
+
+            ex.extract(layer->bottoms[0], bottom_blob);
+            ex.extract(layer->tops[0], top_blob);
+
+            if (bottom_blob.empty() || top_blob.empty())
+            {
+                continue;
+            }
+
+            ncnn::Mat weight_blob(op->kernel_w, op->kernel_h, bottom_blob.c * top_blob.c);
+
+            fprintf(stdout, TEXT_GREEN "Input  [w h nc]: %d %d %d\n" CLR, bottom_blob.w, bottom_blob.h, bottom_blob.c);
+            fprintf(stdout, TEXT_GREEN "Kernel [w h nc]: %d %d %d\n" CLR, op->kernel_w, op->kernel_h, bottom_blob.c * top_blob.c);
+            fprintf(stdout, TEXT_GREEN "Output [w h nc]: %d %d %d\n" CLR, top_blob.w, top_blob.h, top_blob.c);
+
+            // randomize input and kernel
+            gauss_random(bottom_blob);
+            
+            // try every implementation
+            double min_cost = std::numeric_limits<double>::max(); 
+            int best_type = 0;
+
+            // how much conv implementation type ncnn has ?
+            for (int type = 1; type <= 5; ++type)
+            {
+                int support = support_fp32_conv_type(op, bottom_blob, type);
+                if (support < 1)
+                {
+                    // implementation type mismatch convolution configuration, skip
+                    continue;
+                }
+
+                op->impl_type = type;
+
+                auto start = std::chrono::high_resolution_clock::now();
+                const int NREPEATS = 20;
+                op->create_pipeline(opt);
+                for (int repeat = 0; repeat < NREPEATS; ++repeat)
+                {
+                    op->forward(top_blob, bottom_blob, opt);
+                }
+
```

Comment:
though it seems to work properly
it is better calling op->destroy_pipeline(opt); after using this operator